devicetypes/smartthings/zigbee-lock-without-codes/zigbee-lock-without-codes.groovy

```diff
@@ -0,0 +1,257 @@
+				attributeState "locked", label:'locked', action:"lock.unlock", icon:"st.locks.lock.locked", backgroundColor:"#00A0DC", nextState:"unlocking"
+				attributeState "unlocked", label:'unlocked', action:"lock.lock", icon:"st.locks.lock.unlocked", backgroundColor:"#ffffff", nextState:"locking"
+				attributeState "unknown", label:"unknown", action:"lock.lock", icon:"st.locks.lock.unknown", backgroundColor:"#ffffff", nextState:"locking"
+				attributeState "locking", label:'locking', icon:"st.locks.lock.locked", backgroundColor:"#00A0DC"
+				attributeState "unlocking", label:'unlocking', icon:"st.locks.lock.unlocked", backgroundColor:"#ffffff"
+			}
+		}
+		standardTile("lock", "device.lock", inactiveLabel:false, decoration:"flat", width:2, height:2) {
+			state "default", label:'lock', action:"lock.lock", icon:"st.locks.lock.locked", nextState:"locking"
+		}
+		standardTile("unlock", "device.lock", inactiveLabel:false, decoration:"flat", width:2, height:2) {
+			state "default", label:'unlock', action:"lock.unlock", icon:"st.locks.lock.unlocked", nextState:"unlocking"
+
+		}
+		valueTile("battery", "device.battery", inactiveLabel:false, decoration:"flat", width:2, height:2) {
+			state "battery", label:'${currentValue}% battery', unit:""
+		}
+		standardTile("refresh", "device.refresh", inactiveLabel:false, decoration:"flat", width:2, height:2) {
+			state "default", label:'', action:"refresh.refresh", icon:"st.secondary.refresh"
+		}
+
+		main "toggle"
+		details(["toggle", "lock", "unlock", "battery", "refresh"])
+	}
+}
+
+private getCLUSTER_POWER() { 0x0001 }
+private getCLUSTER_DOORLOCK() { 0x0101 }
+private getDOORLOCK_CMD_LOCK_DOOR() { 0x00 }
+private getDOORLOCK_CMD_UNLOCK_DOOR() { 0x01 }
+private getDOORLOCK_RESPONSE_OPERATION_EVENT() { 0x20 }
+private getDOORLOCK_RESPONSE_PROGRAMMING_EVENT() { 0x21 }
+private getPOWER_ATTR_BATTERY_PERCENTAGE_REMAINING() { 0x0021 }
+private getDOORLOCK_ATTR_LOCKSTATE() { 0x0000 }
+
+def installed() {
+	log.debug "Executing installed()"
+	initialize()
+}
+
+def uninstalled() {
+	log.debug "Executing uninstalled()"
+	sendEvent(name:"lockRemoved", value:device.id, isStateChange:true, displayed:false)
+}
+
+def updated() {
+	try {
+		if (!state.init || !state.configured) {
+			state.init = true
+			def cmds = []
+			if (!state.configured) {
+				cmds << initialize()
+			} else {
+				cmds << refresh()
+			}
+
+			return response(cmds.flatten())
+		}
+	} catch (e) {
+		log.warn "ZigBee DTH - updated() threw exception:- $e"
+	}
+	return null
+}
+
+def ping() {
+	refresh()
+}
+
+def refresh() {
+	def cmds = zigbee.readAttribute(CLUSTER_DOORLOCK, DOORLOCK_ATTR_LOCKSTATE)
+	return cmds
+}
+
+def configure() {
+	def cmds = initialize()
+	return cmds
+}
+
+def initialize() {
+	log.debug "Executing initialize()"
+	state.configured = true
+	sendEvent(name:"checkInterval", value:3 * 60 + 2 * 60, displayed:false, data: [protocol:"zigbee", hubHardwareId:device.hub.hardwareID, offlinePingable:"1"])
```

Comment:
How often does the device check in? This value seems low.

Reply:
it looks like you're configuring the lock state and battery to report once an hour, so I think this would result in false "offline"s

Reply:
I changed value of checkInterval to value: 2 * 60 * 60 + 2 * 60.  
Those values (configuration and checkInterval) are taken from current ZigBee lock DTH. 

This device reports battery only when the lock engine is working.